File: 3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        if not nums and len(nums) < 3:
            return []
        result = []
        nums.sort()
        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            self.twoSum(nums, i, result)
        return result
                
    def twoSum(self, nums: List[int], i: int, result: List[List[int]]):
        left = i + 1
        right = len(nums) - 1
        while left < right:
            total = nums[i] + nums[left] + nums[right]
            if total == 0:
                result.append([nums[i], nums[left], nums[right]])
                left += 1
                right -= 1
                
                while left < right and nums[left] == nums[left - 1]:
                    left += 1
            elif total < 0:
                left += 1
            else:
                right -= 1

File: solutions/day_011/3sum.dart
class Solution {
List<List<int>> threeSum(List<int> nums) {
  // Sort the array
  nums.sort();
  // Create a list to store the result
  var res = <List<int>>[];
  // Iterate through the array
  for (var i = 0; i < nums.length; i++) {
    if (i > 0 && nums[i] == nums[i-1]) {
      // Skip the duplicate
      continue;
    }
    var l = i + 1;
    // The right pointer starts from the end of the array
    var r = nums.length - 1;
    // Iterate through the array
    while (l < r) {
      // Calculate the sum
      var s = nums[i] + nums[l] + nums[r];
      // If the sum is less than 0, move the left pointer to the right
      if (s < 0) {
        // Skip the duplicate
        l++;
      // If the sum is greater than 0, move the right pointer to the left
      } else if (s > 0) {
        r--;
      } else {
        // If the sum is 0, add the result to the list
        res.add([nums[i], nums[l], nums[r]]);
        while (l < r && nums[l] == nums[l+1]) {
          l++;
        }
        while (l < r && nums[r] == nums[r-1]) {
          r--;
        }
        // Move the left and right pointers
        l++;
        // Move the right pointer to the left
        r--;
      }
    }
  }
  return res;
}

}